Issue 4433: Ability to define sort order for output of find command

wplapper opened the pull request:
cmd_find: ability to define sort order for output of find command: Add option --reverse

<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------
Added ability to sort output of find command.

<!--
Describe the changes and their purpose here, as detailed as needed.
-->

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->
https://github.com/restic/restic/issues/4433
Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].

Please always follow these steps:
- Read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- Enable [maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- Run `gofmt` on the code in all commits.
- Format all commit messages in the same style as [the other commits in the repository](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
-->

- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

wplapper committed:
Issue 4433:

cmd_find: ability to define sort order for output of find command:
Add option --reverse

wplapper committed:
find: allow sorting of find output

changed wording of changelog/unreleased/issue-4433

MichaelEischer commented on changelog/unreleased/issue-4433:
```diff
@@ -0,0 +1,6 @@
+find: define sorting order of command find.
```
Please follow the format from https://github.com/restic/restic/blob/master/changelog/TEMPLATE

wplapper replied:
done. Also pointed out that the default sorting order has changed.

MichaelEischer commented on cmd/restic/cmd_find.go:
```diff
@@ -73,6 +75,7 @@ func init() {
 	f.BoolVar(&findOptions.PackID, "pack", false, "pattern is a pack-ID")
 	f.BoolVar(&findOptions.ShowPackID, "show-pack-id", false, "display the pack-ID the blobs belong to (with --blob or --tree)")
 	f.BoolVarP(&findOptions.CaseInsensitive, "ignore-case", "i", false, "ignore case for pattern")
+	f.BoolVarP(&findOptions.Reverse, "reverse", "R", false, "reverse sort order newest->oldest")
```
`"reverse sort order from newest to oldest"`

wplapper replied:
Changed to "oldest to newest".

MichaelEischer commented on cmd/restic/cmd_find.go:
```diff
@@ -639,6 +642,9 @@ func runFind(ctx context.Context, opts FindOptions, gopts GlobalOptions, args []
 	sort.Slice(filteredSnapshots, func(i, j int) bool {
 		return filteredSnapshots[i].Time.Before(filteredSnapshots[j].Time)
```
It feels more elegant to let the comparator here check whether `opts.Reverse` is set and directly use the correct order.

wplapper replied:
done.

MichaelEischer commented on cmd/restic/cmd_find.go:
```diff
@@ -73,6 +75,7 @@ func init() {
 	f.BoolVar(&findOptions.PackID, "pack", false, "pattern is a pack-ID")
 	f.BoolVar(&findOptions.ShowPackID, "show-pack-id", false, "display the pack-ID the blobs belong to (with --blob or --tree)")
 	f.BoolVarP(&findOptions.CaseInsensitive, "ignore-case", "i", false, "ignore case for pattern")
+	f.BoolVarP(&findOptions.Reverse, "reverse", "R", false, "reverse sort order newest->oldest")
```
I think we should actually change the default sort order to be from newest to oldest by default. And let the option invert it to the old behavior.

wplapper replied:
done.

MichaelEischer reviewed (changes requested):
Thanks for taking a look at this. The change is simpler than I've initially expected :-) .
A small test case in `cmd/restic/cmd_find_integration_test.go` would also be nice.

As in https://github.com/restic/restic/pull/5183 please rebase the PR instead of adding merge commits.

wplapper commented:
> As in #5183 please rebase the PR instead of adding merge commits.

@MichaelEischer: as I pointed out before, I don't know how to rebase. I am a really beginner when it comes to git. @rawtaz promised me a chat, but I have no idea how I can contact him.

wplapper committed:
restic find: enable sorting order on snapshots

The old sorting behaviour was to sort snapshots from oldest to newest.
The new sorting order is from newest to oldest. If one wants to revert to the
old behaviour, use the option --reverse.

wplapper committed:
restic find - add find integration test

wplapper commented:
Integration test added.

MichaelEischer commented on cmd/restic/cmd_find.go:
```diff
@@ -22,7 +22,10 @@ var cmdFind = &cobra.Command{
 repo.
-It can also be used to search for restic blobs or trees for troubleshooting.`,
+It can also be used to search for restic blobs or trees for troubleshooting.
+The default sort option for the snapshots is youngest to oldest. One can reverse
```
Can be shortened to
```
The default sort option for the snapshots is youngest to oldest. To sort the
output from oldest to youngest specify --reverse.`,
```

MichaelEischer commented on changelog/unreleased/issue-4433:
```diff
@@ -0,0 +1,9 @@
+Enhancement: find: enable sorting of find output, new option --reverse
```
I'd prefer something like the following:
```
Enhancement: Sort `find` output from newest to oldest and add `--reverse` option
```

MichaelEischer commented on cmd/restic/cmd_find_integration_test.go:
```diff
@@ -91,3 +91,53 @@ func TestFindJSON(t *testing.T) {
 	rtest.Assert(t, len(matches[0].Matches) == 3, "expected 3 files to match (%v)", matches[0].Matches)
 	rtest.Assert(t, matches[0].Hits == 3, "expected hits to show 3 matches (%v)", datafile)
 }
+
```
Please run `gofmt` on the file. Ideally, set up your editor to automatically format the file when saving it. For example, in VS Code the Go extension even automatically enables auto formatting of files. Thankfully Go has a canonical formatting style so there's no more need to manually format and indent code.

MichaelEischer commented on cmd/restic/cmd_find_integration_test.go:
```diff
@@ -91,3 +91,53 @@ func TestFindJSON(t *testing.T) {
+
+	// first backup
+	testRunBackup(t, "", []string{env.testdata}, opts, env.gopts)
+	testRunCheck(t, env.gopts)
```
Running `check` is redundant here. (The existing tests call `check` far often for my taste)

MichaelEischer commented on cmd/restic/cmd_find_integration_test.go:
```diff
@@ -91,3 +91,53 @@ func TestFindJSON(t *testing.T) {
+	// first restic find
+	results := testRunFind(t, true, FindOptions{}, env.gopts, "testfile")
+	lines := strings.Split(string(results), "\n")
+	rtest.Assert(t, len(lines) == 2, "expected two files found in repo (%v), found %d", datafile, len(lines))
```
Please also check that find returns the results in the correct order. (The simplest way is probably to call `testListSnapshots(...)` after creating the first snapshot to get the first ID)

MichaelEischer commented on cmd/restic/cmd_find.go:
```diff
@@ -637,7 +642,11 @@ func runFind(ctx context.Context, opts FindOptions, gopts GlobalOptions, args []
 
 	sort.Slice(filteredSnapshots, func(i, j int) bool {
-		return filteredSnapshots[i].Time.Before(filteredSnapshots[j].Time)
+		if opts.Reverse {
```
Our linter rules expect this to be written as follows:
```go
		if opts.Reverse {
			return filteredSnapshots[i].Time.Before(filteredSnapshots[j].Time)
		}
		return filteredSnapshots[i].Time.After(filteredSnapshots[j].Time)
```

You can manually call `golangci-lint run ./...` to get the linting results (or setup your editor accordingly)

MichaelEischer commented on cmd/restic/cmd_find_integration_test.go:
```diff
@@ -91,3 +91,53 @@ func TestFindJSON(t *testing.T) {
 }
+
+
+type testMatches2 struct {
```
Please just use the `testMatches` and `testMatch` that already exist.

MichaelEischer reviewed (changes requested):
Thanks for the test! I'll unfortunately have to bug you with a bit more tooling, this time gofmt (canonical go code formatting) and golangci-lint (checks for code issues), check the comments for more details.

Regarding rebasing, I can recommend the [git book](https://git-scm.com/book/en/v2/Git-Branching-Rebasing), although I can also rebase the PR in the end if necessary. (I can also make the changes mentioned in the review comments if that's fine for you)

wplapper committed:
restic find: implement more recommendations from Michael Eischer

working on the integration test,
rewording the help text for the command

MichaelEischer committed:
make TestFindSorting independent of file order returned by filesystem

MichaelEischer reviewed (approved):
LGTM. Thanks for all the tweaks. I've rebased the PR and added a small bugfix for the test case.
